### PR TITLE
Feature/drive to position command

### PR DESCRIPTION
**Pull request by imathno**
Implemented DriveToPositionCommand
Added new method to DriveTestBase
Added test for DriveToPositionCommand

### CODE REVIEW

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 1]
```diff
@@ -5,6 +5,8 @@
 import com.google.inject.Inject;
 import com.google.inject.Singleton;
 
+import edu.wpi.first.wpilibj.Talon;
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: What's this doing? It shouldn't be needed (if it is, something else is wrong).

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 2]
```diff
@@ -43,6 +49,13 @@ public DriveSubsystem(WPIFactory factory, XPropertyManager propManager) {
         dVelProp = propManager.createPersistentProperty("Drive Vel D", -100);
         fVelProp = propManager.createPersistentProperty("Drive Vel F", 0);
         
+        leftDistanceProp = propManager.createEphemeralProperty("Left Distance", 0);
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Capitalize only the first letter of the property string; everything else should be lowercase.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 3]
```diff
@@ -148,7 +163,6 @@ public void tankDrivePowerMode(double leftPower, double rightPower) {
      * @param motor used to access the motor to set and give data
      */
     private void updateMotorConfig(XCANTalon motor) {
-        motor.configEncoderCodesPerRev((int)encoderCodesProperty.get());
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: Why was this removed? Is the idea to do all the conversions ourselves? If so, the unused property should be removed. Also, the value should be explicitly set to 1 so that we aren't using whatever value was last set.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 4]
```diff
@@ -164,5 +178,20 @@ public void updateMotorTelemetry() {
         leftDriveSlave.updateTelemetryProperties();
         rightDrive.updateTelemetryProperties();
         rightDriveSlave.updateTelemetryProperties();
+        
+        leftDistanceProp.set(leftDrive.getPosition());
+        rightDistanceProp.set(rightDrive.getPosition());
+    }
+    
+    public double getDistance() {
+        return convertToInches(leftDrive.getPosition() - startPosition.get());
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: It'd feel safer if this used both right and left (i.e. averaged them).

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 5]
```diff
@@ -164,5 +178,20 @@ public void updateMotorTelemetry() {
...
+    }
+    
+    public void resetDistance() {
+        startPosition.set(convertToInches(leftDrive.getPosition()));
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: This seems odd; we're using a property to frequently store data for immediate access by other code. I think this value should be stored in a local variable independent of the property.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 4 above)
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: `startPosition` is already in inches, while `getPosition` returns ticks... and then it's put through an inch conversion. Something needs to be moved.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 6]
```diff
@@ -164,5 +178,20 @@ public void updateMotorTelemetry() {
         leftDriveSlave.updateTelemetryProperties();
         rightDrive.updateTelemetryProperties();
         rightDriveSlave.updateTelemetryProperties();
+        
+        leftDistanceProp.set(leftDrive.getPosition());
+        rightDistanceProp.set(rightDrive.getPosition());
+    }
+    
+    public double getDistance() {
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: This isn't distance, this is displacement. It should be named as such.

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    private final PIDManager travelManager;
+    
+    private final DoubleProperty distanceTolerance;
+    private double targetPosition;
+    
+    @Inject
+    public DriveToPositionCommand(DriveSubsystem driveSubsystem, XPropertyManager propManager) {
+        this.driveSubsystem = driveSubsystem;
+        this.requires(driveSubsystem);
+        this.travelManager = new PIDManager("DriveToPosition", propManager, 0.1, 0, 0);
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Spaces and lowercase in string so that it's easier to read.

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        
+        distanceTolerance = propManager.createPersistentProperty("Distance Tolerance", 0.1);
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Wow, that's a pretty tight threshold! This should probably be increased a bit.

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    }
+    
+    public void setTargetPosition(double targetPosition) {
+        this.targetPosition = driveSubsystem.convertToInches(targetPosition);
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: These unit conversions scare me. It doesn't feel natural for the position to be in ticks here (I'd expect everything outside of the drive subsystem to be human units). If we do need to mix units, they should be documented.

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    @Override
+    public void execute() {
+        double power = travelManager.calculate(targetPosition, driveSubsystem.getDistance());
+        
+        driveSubsystem.tankDrivePowerMode(power, power);
+    }
+    
+    @Override
+    public boolean isFinished() {
+        return travelManager.isOnTarget(distanceTolerance.get());
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Again -- are these inches or ticks?

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+import com.google.inject.Inject;
+
+import xbot.common.command.BaseCommand;
+import xbot.common.math.PIDManager;
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.XPropertyManager;
+import competition.subsystems.drive.DriveSubsystem;
+
+public class DriveToPositionCommand extends BaseCommand{
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: This feels more like a "drive for distance" command than a "drive to position" command. A _position_ implies to me that it's a point on the field in both dimensions.

File: Competition2017/tests/competition/subsystems/drive/DriveTestBase.java [hunk 12]
```diff
@@ -30,4 +31,14 @@ public void verifyDriveSetpoints(double left, double right) {
         assertEquals(left, ((MockCANTalon)drive.leftDrive).getSetpoint(), 0.001);
         assertEquals(right, ((MockCANTalon)drive.rightDrive).getSetpoint(), 0.001);
     }
+    
+    public void verifyDrivePositive() {
+        assertTrue(((MockCANTalon)drive.leftDrive).getSetpoint() > 0);
```
Comment (WasabiFan) [inline Competition2017/tests/competition/subsystems/drive/DriveTestBase.java:33]: Is there a comparison assertion method? I can't remember if there is with JUnit.

File: Competition2017/tests/competition/subsystems/drive/commands/DriveToPositionCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        verifyDriveNegative();
+    }
+    
+    @Test
+    public void isFinishedTest() {
+        DriveToPositionCommand command = injector.getInstance(DriveToPositionCommand.class);
+       
+        command.setTargetPosition(2);
+        
+        ((MockCANTalon)drive.leftDrive).setPosition(2);
```
Comment (WasabiFan) [inline Competition2017/tests/competition/subsystems/drive/commands/DriveToPositionCommandTest.java]: Can we also test something within the tolerance but not exactly at the target?

**Review by WasabiFan — DISMISSED** [review]

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 3 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: I think @aschokking told me to remove this because they put in another implementation of this somewhere else.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 4 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: There is no right talon on the summer bot so I was only using the left.

Comment (WasabiFan) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: I'm not sure what you mean. The summer bot has four drive talons, 2 per side, and one encoder per side.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 1 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Didnt mean to import that.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 2 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Doon

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 6 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Done

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 4 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: I was only getting values from the leftDrive

Comment (imathno) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: startPosition starts at ticks if you go up you'll see there is a doubleProperty

Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: oops didn't even notice that.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 3 above)
Comment (WasabiFan) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: What is that other implementation? The inch conversion function?

Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: Idk he just told me to remove this and I think this has nothing to do with the inch conversion function.

Comment (WasabiFan) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: @aschokking can you elaborate?

Comment (aschokking) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: I don't remember anything about this guy, if I had said to remove it that was probably a mistake. 

I agree just leaving it at something like 1.0 would probably be simplest though.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 6 above)
Comment (aschokking) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: These were the debug properties that we should probably remove or rename them to be encoder ticks or something

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 4 above)
Comment (aschokking) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Indeed we were only getting values for one of the encoders and it's really tricky to get at the electronics on the summer bot to verify it was really there on the other side and plugged in etc.. As you said we'll want both sides averaged together on the final bot but I imagine we'll be using the summer bot for drive testing at least for the next few meetings. Maybe let's have the code to handle both there but commented out?

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 5 above)
Comment (aschokking) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: @WasabiFan are you advocating against having a startPosition property or just to also have a local variable in parallel for reading?

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 4 above)
Comment (WasabiFan) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Are you saying that there was a technical problem preventing you from getting data? What was the behavior?

You bring up a good point; it would be nice to be able to operate on just a single encoder if something gets disconnected.

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java (hunk 9 above)
Comment (aschokking) [reply Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Agreed! (should be inches everywhere outside of the drive subsystem)

File: Competition2017/tests/competition/subsystems/drive/DriveTestBase.java (hunk 12 above)
Comment (aschokking) [reply Competition2017/tests/competition/subsystems/drive/DriveTestBase.java:33]: Not that I'm aware of either.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 5 above)
Comment (WasabiFan) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: The latter. I think that the property for telemetry purposes is good; it's just the use of such a property as a local immediate variable that makes me uncomfortable.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 3 above)
Comment (WasabiFan) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: We should check the docs to see if this is a property that is maintained in flash memory. If it is, we need to make sure that we call this everywhere.

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java (hunk 7 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Done

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java (hunk 8 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Updated the value to 1

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java (hunk 9 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Everything in the DriveSubsystem is now in ticks

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java (hunk 10 above)
Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Everything in the DriveToPositionCommand is in inches

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java (hunk 11 above)
Comment (JohnGilb) [reply Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Agree. This should be renamed to something like "DriveForDistanceCommand"

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 14]
```diff
@@ -38,10 +42,17 @@ public DriveSubsystem(WPIFactory factory, XPropertyManager propManager) {
...
-        pVelProp = propManager.createPersistentProperty("Drive Vel P", 2);
-        iVelprop = propManager.createPersistentProperty("Drive Vel I", 0);
-        dVelProp = propManager.createPersistentProperty("Drive Vel D", -100);
-        fVelProp = propManager.createPersistentProperty("Drive Vel F", 0);
+        pVelProp = propManager.createPersistentProperty("Drive vel P", 2);
+        iVelprop = propManager.createPersistentProperty("Drive vel I", 0);
+        dVelProp = propManager.createPersistentProperty("Drive vel D", -100);
+        fVelProp = propManager.createPersistentProperty("Drive vel F", 0);
+        
+        leftDistanceProp = propManager.createEphemeralProperty("Left distance", 0);
```
Comment (aschokking) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: let's rename these guys "Left encoder ticks" instead so we don't get confused about units?

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 15]
```diff
@@ -54,11 +65,13 @@ public DriveSubsystem(WPIFactory factory, XPropertyManager propManager) {
         configMotorTeam(rightDrive, rightDriveSlave);
         rightDrive.createTelemetryProperties("Right master", propManager);
         rightDriveSlave.createTelemetryProperties("Right slave", propManager);
+
+        resetDistance();
```
Comment (aschokking) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Let's add a command to reset the distance as well

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 16]
```diff
@@ -148,7 +161,7 @@ public void tankDrivePowerMode(double leftPower, double rightPower) {
      * @param motor used to access the motor to set and give data
      */
     private void updateMotorConfig(XCANTalon motor) {
-        motor.configEncoderCodesPerRev((int)encoderCodesProperty.get());
+        //motor.configEncoderCodesPerRev((int)encoderCodesProperty.get());
```
Comment (aschokking) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: delete line rather than commenting it out

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java (hunk 10 above)
Comment (JohnGilb) [reply Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Given that we need to be deliberate about this when reading through it, we should probably append "Ticks" or "Inches" to any properties or variables that contain distance information.

so, "distanceToleranceInches"

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 3 above)
Comment (aschokking) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: @JohnGilb is going to verify in the docs if the value is preserved or not.

**Review by JohnGilb — CHANGES_REQUESTED** [review]
See my comments above.

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java (hunk 11 above)
Comment (aschokking) [reply Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: It is 'To' an absolute distance though as opposed to a relative one. I can see your point about position vs distance but it this year sans mechanum we'll probably never have anything about x/y driving so probably won't get confused with anything else.

**Review by aschokking — APPROVED** [review]

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java (hunk 3 above)
Comment (JohnGilb) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java:151]: My most recent read through of the docs suggests that it is not.

Motion Control Profile (MCP) value are saved locally, and those values are:
P Gain: Kp constant to use when control mode is a closed-loop mode.
I Gain: KI constant to use when control mode is a closed-loop mode.
D Gain: KD constant to use when control mode is a closed-loop mode.
F Gain: KF constant to use when control mode is a closed-loop mode.
I Zone: Integral Zone. When nonzero, Integral Accumulator is automatically cleared
when the absolute value of Closed-Loop Error exceeds it.
(Closed-Loop) Ramp Rate: Ramp rate to apply when control mode is a closed-loop
mode.
Allowable Closed-Loop Error: When Closed-Loop Error’s magnitude is less than this
signal, Integral Accum and motor output are auto-zeroed during closed-loop.
Peak Closed-Loop Output: Caps the maximal or peak motor-output during closed-loop.
Nominal Closed-Loop

this does not appear to include encoder configuration.

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 17]
```diff
@@ -164,5 +177,26 @@ public void updateMotorTelemetry() {
...
+        leftDistanceProp.set(leftDrive.getPosition());
+        rightDistanceProp.set(rightDrive.getPosition());
+    }
+    
+    public double getDisplacement() {
+        /* TODO add more values from other motors for accuracy */
+        return convertTicksToInches(leftDrive.getPosition() - startPositionTicks.get());
+    }
+    
+    public void resetDistance() {
```
Comment (aschokking) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: let's call this resetDisplacement so it's clear they're talking about the same thing if we're going with displacement instead

File: Competition2017/src/competition/subsystems/drive/DriveSubsystem.java [hunk 18]
```diff
@@ -47,10 +47,10 @@ public DriveSubsystem(WPIFactory factory, XPropertyManager propManager) {
         dVelProp = propManager.createPersistentProperty("Drive vel D", -100);
         fVelProp = propManager.createPersistentProperty("Drive vel F", 0);
         
-        leftDistanceProp = propManager.createEphemeralProperty("Left distance", 0);
-        rightDistanceProp = propManager.createEphemeralProperty("Right distance", 0);
+        leftEncoderTicksProp = propManager.createEphemeralProperty("Left encoder ticks", 0);
```
Comment (WasabiFan) [inline Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Given that we'll also have left and right shooters, could we include the word "drive" somewhere in the property name?

Comment (imathno) [reply Competition2017/src/competition/subsystems/drive/DriveSubsystem.java]: Done

File: Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    private final PIDManager travelManager;
+    
+    private final DoubleProperty distanceToleranceInches;
+    private double targetPosition;
+    
+    @Inject
+    public DriveToPositionCommand(DriveSubsystem driveSubsystem, XPropertyManager propManager) {
+        this.driveSubsystem = driveSubsystem;
+        this.requires(driveSubsystem);
+        this.travelManager = new PIDManager("Drive to position", propManager, 0.1, 0.1, 0.1);
```
Comment (JohnGilb) [inline Competition2017/src/competition/subsystems/drive/commands/DriveToPositionCommand.java]: Unless you already know that 0.1 is a good constant for the I term, it's recommended to have it be at 0 (since it can lead to extreme robot behavior).

**Review by JohnGilb — APPROVED** [review]
